src/Caching/SqlServer/src/DatabaseOperations.cs

```diff
@@ -259,25 +281,78 @@ public async Task SetCacheItemAsync(string key, byte[] value, DistributedCacheEn
 
             await connection.OpenAsync(token).ConfigureAwait(false);
 
-            using (var reader = await command.ExecuteReaderAsync(
-                CommandBehavior.SequentialAccess | CommandBehavior.SingleRow | CommandBehavior.SingleResult,
-                token).ConfigureAwait(false))
+            if (includeValue)
             {
+                using var reader = await command.ExecuteReaderAsync(
+                    CommandBehavior.SequentialAccess | CommandBehavior.SingleRow | CommandBehavior.SingleResult, token).ConfigureAwait(false);
+
                 if (await reader.ReadAsync(token).ConfigureAwait(false))
                 {
-                    if (includeValue)
+                    if (destination is null)
                     {
                         value = await reader.GetFieldValueAsync<byte[]>(Columns.Indexes.CacheItemValueIndex, token).ConfigureAwait(false);
                     }
+                    else
+                    {
+                        StreamOut(reader, Columns.Indexes.CacheItemValueIndex, destination);
+                        value = []; // use non-null here as a sentinel to say "we got one"
+                    }
                 }
-                else
+            }
+            else
+            {
+                await command.ExecuteNonQueryAsync(token).ConfigureAwait(false);
+            }
+        }
+
+        return value;
+    }
+
+    static long StreamOut(SqlDataReader source, int ordinal, IBufferWriter<byte> destination)
+    {
+        long dataIndex = 0;
+        int read = 0;
+        byte[]? lease = null;
+        do
+        {
+            dataIndex += read; // increment offset
+
+            var memory = destination.GetMemory(8192); // start from the page size
```

Comment:
Nit: magic number to const?

Reply:
done: 552676fbc8b0424ae06c6b2fa748e7f97db7feda